ambry-account/src/main/java/com/github/ambry/account/HelixAccountService.java

```diff
@@ -423,267 +419,176 @@ private void checkOpen() {
-        nameToAccountMap.put(account.getName(), account);
+  void maybePersistNewState(Pair<String, Path> backupPrefixAndPath, Map<String, String> newState) {
+    if (backupPrefixAndPath != null) {
+      try {
+        Path filepath = backupDirPath.resolve(backupPrefixAndPath.getFirst() + NEW_STATE_SUFFIX);
+        writeBackup(filepath, newState);
+      } catch (Exception e) {
+        logger.error("Could not write new state backup file", e);
+        accountServiceMetrics.backupErrorCount.inc();
       }
     }
+  }
 
-    /**
-     * Gets {@link Account} by its id.
-     * @param id The id to get the {@link Account}.
-     * @return The {@link Account} with the given id, or {@code null} if such an {@link Account} does not exist.
-     */
-    private Account getAccountById(Short id) {
-      return idToAccountMap.get(id);
-    }
-
-    /**
-     * Gets {@link Account} by its name.
-     * @param name The id to get the {@link Account}.
-     * @return The {@link Account} with the given name, or {@code null} if such an {@link Account} does not exist.
-     */
-    private Account getAccountByName(String name) {
-      return nameToAccountMap.get(name);
-    }
-
-    /**
-     * Checks if there is an {@link Account} with the given id.
-     * @param id The {@link Account} id to check.
-     * @return {@code true} if such an {@link Account} exists, {@code false} otherwise.
-     */
-    private boolean containsId(Short id) {
-      return idToAccountMap.containsKey(id);
-    }
-
-    /**
-     * Checks if there is an {@link Account} with the given name.
-     * @param name The {@link Account} name to check.
-     * @return {@code true} if such an {@link Account} exists, {@code false} otherwise.
-     */
-    private boolean containsName(String name) {
-      return nameToAccountMap.containsKey(name);
+  /**
+   * Save the zookeeper state from before the update to disk.
+   * The following file name format will be used: {@code {yyyyMMdd}T{HHmmss}.{unique long}.old}.
+   * If there are multiple files with the same timestamp, the unique long will prevent the file names from clashing.
+   */
+  void maybePersistOldState(Pair<String, Path> backupPrefixAndPath, Map<String, String> oldState) {
+    if (backupPrefixAndPath != null) {
+      try {
+        writeBackup(backupPrefixAndPath.getSecond(), oldState);
+      } catch (Exception e) {
+        logger.error("Could not write previous state backup file", e);
+        accountServiceMetrics.backupErrorCount.inc();
+      }
     }
+  }
 
-    /**
-     * Gets all the {@link Account}s in this {@code AccountInfoMap} in a {@link Collection}.
-     * @return A {@link Collection} of all the {@link Account}s in this map.
-     */
-    private Collection<Account> getAccounts() {
-      return Collections.unmodifiableCollection(idToAccountMap.values());
+  /**
+   * If a backup file has not yet been created, reserve a new one with the following file name format:
```

Comment:
Minor: not quite accurate since you are reserving a new backup file regardless of whether one already exists.

Reply:
updated